sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala

```diff
@@ -65,13 +65,38 @@ class JDBCOptions(
   // Required parameters
   // ------------------------------------------------------------
   require(parameters.isDefinedAt(JDBC_URL), s"Option '$JDBC_URL' is required.")
-  require(parameters.isDefinedAt(JDBC_TABLE_NAME), s"Option '$JDBC_TABLE_NAME' is required.")
+
   // a JDBC URL
   val url = parameters(JDBC_URL)
-  // name of table
-  val table = parameters(JDBC_TABLE_NAME)
+  val tableName = parameters.get(JDBC_TABLE_NAME)
+  val query = parameters.get(JDBC_QUERY_STRING)
+  // Following two conditions make sure that :
+  // 1. One of the option (dbtable or query) must be specified.
+  // 2. Both of them can not be specified at the same time as they are conflicting in nature.
+  require(
+    tableName.isDefined || query.isDefined,
+    s"Option '$JDBC_TABLE_NAME' or '${JDBC_QUERY_STRING}' is required."
+  )
+
+  require(
+    !(tableName.isDefined && query.isDefined),
+    s"Both '$JDBC_TABLE_NAME' and '$JDBC_QUERY_STRING' can not be specified."
+  )
+
+  // table name or a table expression.
+  val tableExpression = tableName.map(_.trim).getOrElse {
+    // We have ensured in the code above that either dbtable or query is specified.
+    query.get match {
+      case subq if subq.nonEmpty => s"(${subq}) spark_gen_${curId.getAndIncrement()}"
```

Comment:
We need `curId`? A constant name is bad?

Reply:
@maropu Don't mind using a constant name ? "spark_gen_alias" ?

Reply:
yea, simpler, better. Btw, we essentially need the alias name for a query? When we currently describe  a query in `dbtable`, it seems we don't need the name? https://github.com/apache/spark/blob/bc111463a766a5619966a282fbe0fec991088ceb/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala#L1196

Reply:
@maropu Yeah. we need an alias. Systems like postgress require a mandatory table subquery alias.